### PR TITLE
Enable `-preview=fixAliasThis`

### PR DESCRIPTION
**Pull request by thewilsonator**
... turn the old behaviour into a `-revert=` option.

rebases #15436

### CODE REVIEW

**Comment by thewilsonator** [comment]
The heck is wrong with Circle here?

**Comment by thewilsonator** [comment]
... https://github.com/dlang/dmd/pull/15947 how did that not fail spectacularly well before this PR???

**Comment by thewilsonator** [comment]
Buildkite failures:

- [ ]  `snazzy-d/sdc`: [src/source/location.d(230,51): Error: undefined identifier `isMixin`](https://github.com/snazzy-d/sdc/blob/master/src/source/location.d#L168)
- [ ] `ldc-developers/ldc`: same frontend issue as https://github.com/dlang/dmd/pull/15947
- [ ] `atilaneves/unit-threaded`: some forward referencing error.

**Comment by thewilsonator** [comment]
@kinke what's the best way to get https://github.com/dlang/dmd/pull/15947 into LDC for this PR: cherry pick it? wait until we merge upstream?

cc @ibuclaw this will affect GDC too

**Comment by ibuclaw** [comment]
> cc @ibuclaw this will affect GDC too

Don't think so. If I understand the ldc failure correctly, you're using a newer version of the compiler to build an older version. As a general rule of thumb, don't do that.

[As I document](https://gcc.gnu.org/install/prerequisites.html):

> More recent versions of GDC than the version built are not guaranteed to work and will often fail during the build with compilation errors relating to deprecations or removed features.

**Comment by kinke** [comment]
> ... #15947 how did that not fail spectacularly well before this PR???

Oh wow, this is really wild! So as soon as this PR lands, compilers based on that new frontend aren't able to compile older frontends (v2.106 and older) anymore. I guess that's a general problem, e.g., for bisecting compiler regressions using a new host compiler, and the current Buildkite problem for latest LDC stable just one symptom of it.

File: compiler/test/compilable/test5973.d [hunk 1]
```diff
@@ -25,7 +25,7 @@ class F : E
 {
     void test()
     {
-        i = 5;
```
Comment (dkorpel) [inline compiler/test/compilable/test5973.d:28]: `fixAliasThis` is supposed to change the search order so ambiguities are resolved differently. It's not supposed to turn single `alias this` hits into errors, however weird the `sizeof.max` turning into `typeof(this.extent.sizeof).max` may be. The failure in dmd + buildkite projects suggests the implementation bugged and the switch is not ready to become the default yet.

Comment (WalterBright) [reply compiler/test/compilable/test5973.d:28]: I thought we agreed that we weren't going to fix alias this, that it was unfixable, and we should just stick with the way it is now? (The reason is that fixing it just breaks something else.)

Comment (thewilsonator) [reply compiler/test/compilable/test5973.d:28]: FWIW this appears to only affect `alias this` hits within the `struct/(sub)class` that defines it and only when the hit is implicitly `[this.]foo` as opposed to `other.foo` as in [`auto getSource() out(result; result.isMixin() == isMixin())`](https://github.com/snazzy-d/sdc/blob/master/src/source/location.d#L168), the `result.isMixin()` is fine, the naked `isMixin()` is apparently not.

Also,
```d
new F().i;
```
would work perfectly fine here. Why this happens I do not know and defer to @RazvanN7 . If that were to be fixed (which on the surface seems to be an orthogonal issue) then this particular change would not need to be made.

**Comment by RazvanN7** [comment]
Closing this for now until we decide what we do next with alias this.

**Pull request closed without merging**